test(ivy): create todo app in ivy

mhevery opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
test(ivy): create todo app in ivy

kara commented on packages/core/test/bundling/todo/index.html:
```diff
@@ -0,0 +1,31 @@
+      - `bundle.min.js`: Is what the site would serve to their users. It has gone
+        through rollup, build-optimizer, and uglify with tree shaking.
+      - `bundle.min_debug.js`: Is what the developer would like to see when debugging
+        the application. It has also done through full pipeline of rollup, build-optimizer,
```
done -> gone

mhevery replied:
fixed

kara reviewed (approved):
LGTM

mhevery commented on tools/public_api_guard/core/core.d.ts:
```diff
@@ -370,7 +370,8 @@ export declare const HostListener: HostListenerDecorator;
 /** @experimental */
 export declare function inject<T>(token: Type<T> | InjectionToken<T>, notFoundValue?: undefined, flags?: InjectFlags): T;
-export declare function inject<T>(token: Type<T> | InjectionToken<T>, notFoundValue: T | null, flags?: InjectFlags): T | null;
+export declare function inject<T>(token: Type<T> | InjectionToken<T>, notFoundValue: T, flags?: InjectFlags): T;
```
This change is needed to otherwise `inject(Foo, new Foo())`'s type is `Foo|null` which is not correct.

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -762,7 +751,7 @@ export function createFactory(
     }
     if (dependency.isHost) {
-      unsupported('host dependencies');
+      // unsupported('host dependencies');
```
why ? add a comment ?

mhevery replied:
Opps, will revert

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -599,17 +599,6 @@ class TemplateDefinitionBuilder implements TemplateAstVisitor, LocalResolver {
             this._bindingMode, directive.sourceSpan, R3.elementProperty, o.literal(nodeIndex),
             o.literal(input.templateName), o.importExpr(R3.bind).callFn([convertedBinding]));
       }
-
```
Should those changes be part of a different commit describing the rationale for the change ?

mhevery replied:
broken to separate commits

vicb commented on packages/core/test/bundling/todo/BUILD.bazel:
```diff
@@ -0,0 +1,68 @@
+    # path, to disambiguate from other workspaces.
+    # Here, the rule implementation is looking in an execroot where the layout
+    # has an "external" directory for external dependencies.
+    # This should probably start with "angular/" and let the rule deal with it.
```
`# TODO: Remove once #22913 lands` ?

mhevery replied:
https://github.com/angular/angular/pull/22913 is unrelated to this.

vicb replied:
Well... that comes from you: https://github.com/angular/angular/pull/22654#discussion_r176233174

What did you mean then ?

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -0,0 +1,120 @@
+    {text: 'Demonstrate Components', done: false},
+    {text: 'Demonstrate Structural Directives', done: true},
+    {text: 'Demonstrate NgModules', done: false},
+    {text: 'Demonstrate zoneless changed detection', done: false},
```
typo "change"

mhevery replied:
fixed

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -0,0 +1,120 @@
+  // The reason for the guard is that template executes creation and binding together
+  // but NgForOf expects creation and binding separate.
+  template: `
+    <div [class.done]="todo && todo.done">
```
`todo?.done` ?

mhevery replied:
I intentionally left it in this ugly state because it is a hack which should be removed once we fix it in ivy runtime.

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -0,0 +1,120 @@
+
+  // TODO(misko): Injection is broken because the compiler generates incorrect code.
+  constructor(/**appState: AppState*/) {
+    // TODO(misko): remove once injection is working.
```
duplicate TODO ?

mhevery replied:
fixed

IgorMinar reviewed (approved):
lgtm for api changes

but I'm surprised to see that we now have ivy_ng_module - fragmenting the build system like this seem like a bad idea. did you consult this with @alexeagle ?

mhevery commented:
@IgorMinar Yes it is a bad idea. 🙂  It is internal only build rule/hack until we have proper rule https://github.com/angular/angular/blob/master/packages/compiler/design/separate_compilation.md#ng_module-output-bazel and yes this was discussed with @alexeagle @alxhub and @chuckjaz

mhevery committed:
ci: update to latest rulse_nodejs

mhevery committed:
refactor(ivy): cleanup necessary domino files

mhevery committed:
feat(ivy): add `(event)="handle"` syntax support to compiler

mhevery commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -0,0 +1,120 @@
+  `
+})
+export class TodoComponent {
+  static DEFAULT_TODO: ToDo = {text: '', done: false};
```
removed

mhevery committed:
refactor(ivy): align compiler with runtime

Remove `containerRefreshStart` and `containerRefreshEnd` instruction
from the output.

Generate directives as a list in `componentDef` rather than inline into
instructions. This is consistent in making selector resolution runtime
so that translation of templates can follow locality.

kara reviewed (approved):
LGTM

trotyl commented on packages/compiler/test/render3/r3_compiler_compliance_spec.ts:
```diff
@@ -211,23 +207,19 @@ describe('compiler compliance', () => {
-              $r3$.ɵE(0, 'ul', null, null, $c1$);
-              $r3$.ɵC(2, $c2$, MyComponent_IfDirective_Template_2);
+              $r3$.ɵE(0, 'ul', null, $c1$);
+              $r3$.ɵC(2, MyComponent_IfDirective_Template_2, null, $c2$);
```
Since `IfDirective` is not determined at this stage any more, how could it still remains in the template function name? Same for `MyComponent_ForOfDirective_Template_1` below.

mhevery replied:
You are right it should not be in the function name. We currently are still using old backend for the compiler which still has this information. The new compiler pipeline `ngtsc` will not and so when we switch over we will not be able to generate the string. Since the function name is there just to aid in debugging, i am going to leave it for now since it will get fixe automatically when we switch over to `ngtsc`.

trotyl commented on packages/core/src/render3/instructions.ts:
```diff
@@ -400,7 +400,8 @@ function resetApplicationState() {
  * @param host The host element node to use
- * @param defs Any directive or pipe defs that should be used for matching
+ * @param directives Any directive defs that should be used for matching
+ * @param pipes Any pipe defs that should be used for matching
```
`Any directive defs` -> `Any directive types` according to the change? (Should `any` be combined with plural?)

mhevery replied:
Removed the word `any`. Yes it is deys not types. We pass types to `defineComponent` which in turn converts them to defs. Because going from Type to Defs is megamorphic we want to do it as early as possible and than have all of the runtime be defs only.
